### PR TITLE
Feat: Social Network API

### PR DESCRIPTION
**Pull request by zendy1994**

### CODE REVIEW

**Review by michael-do-goldenowl — COMMENTED** [review]
why having both pnpm and yarn

File: src/database/migrations/1706692679617-UserFollowMigration.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    await queryRunner.createTable(
+      new Table({
+        name: TableDB.USER_FOLLOW,
+        columns: [
+          {
+            name: "id",
+            type: "uuid",
+            isPrimary: true,
+            generationStrategy: "uuid",
+            default: `uuid_generate_v4()`,
```
Comment (michael-do-goldenowl) [inline src/database/migrations/1706692679617-UserFollowMigration.ts]: should have a migration helper file which has some helper for id generation, timestamp to reuse b.c they are used in many where

File: src/database/migrations/1706692679617-UserFollowMigration.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            type: "uuid",
+          },
+          {
+            name: "following_id",
+            type: "uuid",
+          },
+          {
+            name: "created_at",
+            type: "timestamptz",
+            default: "now()",
```
Comment (michael-do-goldenowl) [inline src/database/migrations/1706692679617-UserFollowMigration.ts]: same above

File: src/database/migrations/1706692679617-UserFollowMigration.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        referencedColumnNames: ["id"],
+        referencedTableName: TableDB.USER,
+        onDelete: "CASCADE",
+      })
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    const table = await queryRunner.getTable(TableDB.USER_FOLLOW);
+    const followingForeignKey = table.foreignKeys.find(
```
Comment (michael-do-goldenowl) [inline src/database/migrations/1706692679617-UserFollowMigration.ts]: same here, should create a helper fucntion to shortcut drop foreign key

File: src/app/auth/guards/jwt-auth.guard.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  handleRequest(err: any, user: any, _info: any, context: any, _status: any) {
+    if (err || !user || this.isTokenBlacklisted(user, context)) {
+      throw err || new UnauthorizedException("Token has expired");
+    }
+
+    return user;
+  }
+
+  private isTokenBlacklisted(user: any, context: any): boolean {
```
Comment (michael-do-goldenowl) [inline src/app/auth/guards/jwt-auth.guard.ts]: should use exact type instead of any

File: src/app/auth/guards/jwt-auth.guard.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  constructor(private readonly reflector: Reflector) {
+    super();
+  }
+
+  canActivate(context: ExecutionContext) {
+    return super.canActivate(context);
+  }
+
+  handleRequest(err: any, user: any, _info: any, context: any, _status: any) {
+    if (err || !user || this.isTokenBlacklisted(user, context)) {
```
Comment (michael-do-goldenowl) [inline src/app/auth/guards/jwt-auth.guard.ts]: if error throw error right now for cleaner condition and also this function has many params then should use one object as params for clean

File: src/app/auth/interfaces/auth.interfaces.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export interface IJwtPayload {
```
Comment (michael-do-goldenowl) [inline src/app/auth/interfaces/auth.interfaces.ts:1]: this file should be under AuthModule

File: src/app/file/file.controller.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+@ApiTags('File')
+@Controller('file')
+@UseGuards(JwtAuthGuard)
+export class FileController {
+  constructor(private fileService: FileService) { }
+
+  @Post('/upload')
+  @UseInterceptors(FileInterceptor('file'))
+  uploadPublicFile(@UploadedFile() file: Express.Multer.File): Promise<File> {
```
Comment (michael-do-goldenowl) [inline src/app/file/file.controller.ts:25]: should have file type and file size validation

File: src/app/follow/dto/create-follow.dto.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { IsNotEmpty, IsUUID } from "class-validator";
+
+export class CreateFollowDto {
+  @IsNotEmpty()
+  @IsUUID()
+  follow_user_id: string;
```
Comment (michael-do-goldenowl) [inline src/app/follow/dto/create-follow.dto.ts]: should be camel case

File: src/database/sql/database.module.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import { Module } from "@nestjs/common";
+import { ConfigService } from "@nestjs/config";
+import { TypeOrmModule } from "@nestjs/typeorm";
+
+@Module({
```
Comment (michael-do-goldenowl) [inline src/database/sql/database.module.ts]: DatabaseModule should be outside sql folder, sql folder only for file with .sql

File: src/interceptors/transform.interceptor.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import {
+  CallHandler,
+  ExecutionContext,
+  Injectable,
+  NestInterceptor,
+} from "@nestjs/common";
+import { instanceToPlain } from "class-transformer";
+import { map } from "rxjs/operators";
+
+@Injectable()
```
Comment (michael-do-goldenowl) [inline src/interceptors/transform.interceptor.ts:10]: what is it for?

File: src/interceptors/transform.interceptor.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+export class TransformInterceptor implements NestInterceptor {
```
Comment (michael-do-goldenowl) [inline src/interceptors/transform.interceptor.ts:6]: It is used for custom exception filter right? As nestjs docs almost case we dont need to custom it, let use default standard nestjs exception
https://docs.nestjs.com/exception-filters#custom-exceptions

File: src/main.ts [hunk 12]
```diff
@@ -1,8 +1,61 @@
...
+import { NestFactory } from "@nestjs/core";
+import { DocumentBuilder, SwaggerModule } from "@nestjs/swagger";
+import * as Sentry from "@sentry/node";
+import { config } from "aws-sdk";
+import dayjs from "dayjs";
+import timezone from "dayjs/plugin/timezone";
+import utc from "dayjs/plugin/utc";
+
+dayjs.extend(utc);
+dayjs.extend(timezone);
```
Comment (michael-do-goldenowl) [inline src/main.ts]: should create another file for dayjs and import it in here for centralization

File: src/main.ts [hunk 13]
```diff
@@ -1,8 +1,61 @@
...
+  app.useGlobalPipes(new ValidationPipe());
+  app.useGlobalInterceptors(new TransformInterceptor());
+
+  app.setGlobalPrefix("api/v1", { exclude: ["/"] });
+
+  config.update({
+    accessKeyId: configService.get<string>("AWS_ACCESS_KEY_ID"),
+    secretAccessKey: configService.get<string>("AWS_SECRET_ACCESS_KEY"),
+    region: configService.get<string>("AWS_REGION"),
+  });
```
Comment (michael-do-goldenowl) [inline src/main.ts]: why need to put it in here while can move into file module

File: src/schema/config.schema.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    DATABASE_URL: Joi.string().required(),
+    SENTRY_DNS: Joi.string().required(),
+    JWT_SECRET_KEY: Joi.string().required(),
+    JWT_EXPIRES_TIME: Joi.string().required(),
+    AWS_REGION: Joi.string().required(),
+    AWS_ACCESS_KEY_ID: Joi.string().required(),
+    AWS_SECRET_ACCESS_KEY: Joi.string().required(),
+    AWS_PUBLIC_BUCKET_NAME: Joi.string().required(),
+    EMAIL_VERIFY_OTP: Joi.string().required(),
+    EMAIL_PASSWORD_VERIFY_OTP: Joi.string().required(),
```
Comment (michael-do-goldenowl) [inline src/schema/config.schema.ts]: they have some common config, like EMAIL_VERIFY_OTP and EMAIL_PASSWORD_VERIFY_OTP so let create a common config first

**Comment by michael-do-goldenowl** [comment]
Also should capture db diagram to easy to see

File: src/app/follow/follow.module.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { Module } from "@nestjs/common";
+import { FollowService } from "./follow.service";
+import { FollowController } from "./follow.controller";
+import { TypeOrmModule } from "@nestjs/typeorm";
+import { UserModule } from "../user/user.module";
+import { User } from "../user/entities/user.entity";
+
+@Module({
+  imports: [TypeOrmModule.forFeature([User]), UserModule],
```
Comment (michael-do-goldenowl) [inline src/app/follow/follow.module.ts]: TypeOrmModule.forFeature([User] is every where, while it should be only in user module, let research about repository-service pattern and create a file name user.repository.ts

File: src/app/user/user.controller.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  Patch,
+  Post,
+  Query,
+  UploadedFile,
+  UseGuards,
+  UseInterceptors,
+} from "@nestjs/common";
+import { FileInterceptor } from "@nestjs/platform-express";
+
+@Controller("user")
```
Comment (michael-do-goldenowl) [inline src/app/user/user.controller.ts]: should have a constant file for all route in app
st like
Routes = {
    user: '/user',
    file: '/file'
}

File: src/app/user/user.controller.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  @UseGuards(JwtAuthGuard)
+  @UseInterceptors(
+    FileInterceptor("avatar"),
+    new PublicFileValidatorInterceptor(
+      [/^image\/(jpg|jpeg|png|gif|webp)$/i],
+      "Only JPG, JPEG, PNG, GIF and WEBP file are allowed.",
+      false
+    )
+  )
+  async addAvatar(
```
Comment (michael-do-goldenowl) [inline src/app/user/user.controller.ts:43]: move this fn to service, at controller level we only handle request/response

**Comment by zendy1994** [comment]
Screenshot 2024-02-19 at 14 17 12

**Pull request merged**